### ISSUE TITLE
Explicit self type ignored for inherited constructor

### ISSUE THREAD

**Issue by comex** [python/mypy#14055]
**Bug Report**

In the following code, it should be impossible to construct `Sub`, because it inherits from `Super[int]`, but `Super`'s `__init__` function requires `self: 'Super[bytes]'`.  But mypy doesn't report any error.

**To Reproduce**

```python
from typing import Generic, TypeVar

T = TypeVar('T')

class Super(Generic[T]):
    def __init__(self: 'Super[bytes]', foo: bool) -> None:
        pass

class Sub(Super[int]):
    pass

x = Sub(foo=True)
```
https://mypy-play.net/?mypy=latest&python=3.10&gist=d747064b8573ba4b6ed3dc7850fab8e7

(same behavior with `--strict`)

**Expected Behavior**

Type error

**Actual Behavior**

No error

**Your Environment**

Playground, `--strict`

**Comment by AlexWaygood** [comment]
This feels slightly contrived. If you want to forbid `T` from being solved as a type that isn't a subtype of `bytes`, the way to to do that is to use the `bound` keyword. Mypy correctly emits an error on this snippet:

```python
from typing import Generic, TypeVar

T = TypeVar('T', bound=bytes)

class Super(Generic[T]):
    def __init__(self, foo: bool) -> None:
        pass

class Sub(Super[int]):  # error: Type argument "int" of "Super" must be a subtype of bytes
    pass
```

https://mypy-play.net/?mypy=latest&python=3.10&gist=291b0390fb1c68dd8904e12b3e81a39e

**Comment by DevilXD** [comment]
@comex Your reproduction code makes no sense - you're defining a class as a `Generic[T]`, but then never use the `T` in the rest of the definition. Your `Super[bytes]` self type doesn't (and cannot) cause `T` to "infer the type backwards", because that's not how it was ever supposed to work. If you want to restrict `T` to only be `bytes` or it's subclass, you need to define the type var itself as such, per what @AlexWaygood said.

**Comment by AlexWaygood** [comment]
I'm going to go ahead and close this for now :)
